src/kvdb/prometheus.rs

```diff
@@ -0,0 +1,135 @@
+pub struct KvdbPrometheusOptions {
+	/// Prometheus endpoint port.
+	#[clap(long, default_value = "65432")]
+	port: u16,
+	/// Database poll timeout (default, once per 5 minutes).
+	#[clap(long, default_value = "300.0")]
+	poll_timeout: f32,
+}
+
+struct KvdbPrometheusMetrics {
+	keys_size_gauge: IntGaugeVec,
+	values_size_gauge: IntGaugeVec,
+	elements_count_gauge: IntGaugeVec,
+}
+
+pub async fn run_prometheus_endpoint_with_db<D: IntrospectorKvdb + Send + Sync + 'static>(
+	db: D,
+	prometheus_opts: KvdbPrometheusOptions,
+) -> Result<Vec<tokio::task::JoinHandle<()>>> {
+	let prometheus_registry = Registry::new_custom(Some("introspector".into()), None)?;
+	let metrics = Arc::new(Mutex::new(register_metrics(&prometheus_registry)));
+	let socket_addr =
+		std::net::SocketAddr::new(std::net::IpAddr::V4(std::net::Ipv4Addr::new(0, 0, 0, 0)), prometheus_opts.port);
+	let mut futures: Vec<tokio::task::JoinHandle<()>> = vec![];
+	futures.push(tokio::spawn(async move {
+		prometheus_endpoint::init_prometheus(socket_addr, prometheus_registry)
+			.await
+			.unwrap()
+	}));
+	futures.push(tokio::spawn(async move {
+		update_db(db, metrics.clone(), prometheus_opts).await;
+	}));
+	info!("Starting prometheus node on {:?}", socket_addr);
+
+	Ok(futures)
+}
+
+struct UpdateResult {
+	column: String,
+	keys_count: i64,
+	keys_space: i64,
+	values_space: i64,
+}
+
+async fn update_db<D: IntrospectorKvdb>(
+	db: D,
+	metrics: Arc<Mutex<KvdbPrometheusMetrics>>,
+	prometheus_opts: KvdbPrometheusOptions,
+) {
+	loop {
+		info!("Starting update db iteration");
+		let columns = db.list_columns().unwrap();
+		let mut update_results: Vec<UpdateResult> = vec![];
+		for col in columns {
+			let mut keys_space = 0_i64;
+			let mut keys_count = 0_i64;
+			let mut values_space = 0_i64;
+
+			info!("Iterating over column {}", col.as_str());
+			let iter = db.iter_values(col.as_str()).unwrap();
+			for (key, value) in iter {
+				keys_space += key.len() as i64;
+				keys_count += 1;
+				values_space += value.len() as i64;
+			}
+
+			update_results.push(UpdateResult { column: col.clone(), keys_count, keys_space, values_space });
+		}
+
+		let unlocked_metrics = metrics.lock().await;
+
+		for res in &update_results {
+			unlocked_metrics
+				.elements_count_gauge
+				.with_label_values(&[res.column.as_str()])
+				.set(res.keys_count);
+			unlocked_metrics
+				.keys_size_gauge
+				.with_label_values(&[res.column.as_str()])
+				.set(res.keys_space);
+			unlocked_metrics
+				.values_size_gauge
+				.with_label_values(&[res.column.as_str()])
+				.set(res.values_space);
+		}
+		info!("Ended update db iteration");
+		tokio::time::sleep(std::time::Duration::from_secs_f32(prometheus_opts.poll_timeout)).await;
+	}
+}
+
+fn register_metrics(registry: &Registry) -> KvdbPrometheusMetrics {
+	let elements_count_gauge = prometheus_endpoint::register(
+		IntGaugeVec::new(Opts::new("kvdb_elements_count", "Number of keys in kvdb"), &["column"]).unwrap(),
```

Comment:
We could use expect for more explicit errors.

Reply:
I doubt that is useful as I cannot even imagine conditions why this could fail :)